multitenant: TestTenantUpgradeFailure/upgrade_tenant_have_it_crash_then_resume flaked on master

See flake [here ](https://teamcity.cockroachdb.com/viewLog.html?buildId=10804367&buildTypeId=Cockroach_BazelEssentialCi)

Jira issue: CRDB-29478
Epic: CRDB-18499

Documenting failure modes observed so far:
- initial [report](https://github.com/cockroachdb/cockroach/issues/106279#issue-1791649849) from Michael:
```
Error Trace:	pkg/ccl/kvccl/kvtenantccl/upgradeccl/upgradeccl_test_test/pkg/ccl/kvccl/kvtenantccl/upgradeccl/tenant_upgrade_test.go:244
        	            				pkg/ccl/kvccl/kvtenantccl/upgradeccl/upgradeccl_test_test/pkg/ccl/kvccl/kvtenantccl/upgradeccl/tenant_upgrade_test.go:347
        	Error:      	Received unexpected error:
        	            	migration-job-find-already-completed: failed to connect to n1 at 127.0.0.1:42029: initial connection heartbeat failed: grpc: connection error: desc = "error reading server preface: http2: frame too large" [code 14/Unavailable]
```
- [reproduction](https://github.com/cockroachdb/cockroach/issues/106279#issuecomment-1626071406) from Ahmad:
```
I230707 20:26:25.928348 22 testutils/testcluster/testcluster.go:145  [-] 5682  TestCluster quiescing nodes
     tenant_upgrade_test.go:340: expected query 'SET CLUSTER SETTING version = $1' error '.*(database is closed|failed to connect|closed network connection|upgrade      failed due to transient SQL servers)+', got: pq: migration-job-mark-job succeeded: tenant-side-limiter pool closed: shutting down
```
- [reproduction](https://github.com/cockroachdb/cockroach/issues/106279#issuecomment-1630582198) from Tobi:
```
    sql_runner.go:313: error executing 'SELECT count(*) FROM system.sql_instances WHERE session_id IS NOT NULL': pq: failed to connect to n1 at 127.0.0.1:37095: initial connection heartbeat failed: grpc: connection error: desc = "error reading server preface: http2: frame too large" [code 14/Unavailable]
```
- [2nd reproduction](https://github.com/cockroachdb/cockroach/issues/106279#issuecomment-1631068841) from Michael:
```
    sql_runner.go:313: error executing 'SELECT count(*) FROM system.sql_instances WHERE session_id IS NOT NULL': pq: failed to connect to n2 at 127.0.0.1:37551: initial connection heartbeat failed: grpc: client cluster ID "fbc7e75d-63c9-44f4-8124-c3efdf9d5265" doesn't match server cluster ID "7d710a5e-a37a-43ff-ac9a-ace152e5b8f0" [code 2/Unknown]
```

--- a/pkg/ccl/kvccl/kvtenantccl/upgradeccl/BUILD.bazel
+++ b/pkg/ccl/kvccl/kvtenantccl/upgradeccl/BUILD.bazel
@@ -22,7 +22,6 @@ go_test(
         "//pkg/sql/sqlinstance/instancestorage",
         "//pkg/sql/sqlliveness/slinstance",
         "//pkg/testutils/serverutils",
-        "//pkg/testutils/skip",
         "//pkg/testutils/sqlutils",
         "//pkg/upgrade",
         "//pkg/upgrade/upgradebase",

--- a/pkg/ccl/kvccl/kvtenantccl/upgradeccl/tenant_upgrade_test.go
+++ b/pkg/ccl/kvccl/kvtenantccl/upgradeccl/tenant_upgrade_test.go
@@ -24,7 +24,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlinstance/instancestorage"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slinstance"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/upgrade"
 	"github.com/cockroachdb/cockroach/pkg/upgrade/upgradebase"
@@ -222,6 +221,7 @@ func TestTenantUpgradeFailure(t *testing.T) {
 
 	// Channel for stopping a tenant.
 	tenantStopperChannel := make(chan struct{})
+	shouldWaitForTenantToStop := true
 	startAndConnectToTenant := func(t *testing.T, id uint64) (tenant serverutils.ApplicationLayerInterface, db *gosql.DB) {
 		settings := cluster.MakeTestingClusterSettingsWithVersions(
 			v2,
@@ -268,8 +268,19 @@ func TestTenantUpgradeFailure(t *testing.T) {
 									ctx context.Context, version clusterversion.ClusterVersion, deps upgrade.TenantDeps,
 								) error {
 									t.Log("v2 migration starts running")
-									tenantStopperChannel <- struct{}{}
-									t.Log("v2 migration finishes running")
+									defer t.Log("v2 migration finishes running")
+									if shouldWaitForTenantToStop {
+										tenantStopperChannel <- struct{}{}
+										// Wait until we are sure the stopper is quiescing.
+										for {
+											select {
+											case <-tenant.Stopper().ShouldQuiesce():
+												return nil
+											default:
+												continue
+											}
+										}
+									}
 									return nil
 								}), true
 						default:
@@ -287,7 +298,6 @@ func TestTenantUpgradeFailure(t *testing.T) {
 	}
 
 	t.Run("upgrade tenant have it crash then resume", func(t *testing.T) {
-		skip.WithIssue(t, 106279)
 		t.Log("create a tenant before upgrading anything")
 		const initialTenantID = 10
 		tenant, conn := startAndConnectToTenant(t, initialTenantID)
@@ -316,8 +326,7 @@ func TestTenantUpgradeFailure(t *testing.T) {
 		db.CheckQueryResults(t,
 			"SELECT * FROM t", [][]string{{"1"}, {"2"}})
 		t.Log("upgrade the tenant cluster, expecting the upgrade to fail on v1")
-		db.ExpectErr(t,
-			".*(database is closed|failed to connect|closed network connection|upgrade failed due to transient SQL servers)+",
+		db.ExpectNonNilErr(t,
 			"SET CLUSTER SETTING version = $1", v2.String())
 
 		t.Log("waiting for tenant shutdown to complete")
@@ -361,6 +370,7 @@ func TestTenantUpgradeFailure(t *testing.T) {
 			[][]string{{"1"}})
 
 		t.Log("upgrade the tenant cluster to v2")
+		shouldWaitForTenantToStop = false
 		db.Exec(t, "SET CLUSTER SETTING version = $1", v2.String())
 
 		close(tenantStopperChannel)

--- a/pkg/testutils/sqlutils/sql_runner.go
+++ b/pkg/testutils/sqlutils/sql_runner.go
@@ -151,6 +151,15 @@ func (sr *SQLRunner) ExpectErr(t Fataler, errRE string, query string, args ...in
 	sr.expectErr(t, query, err, errRE)
 }
 
+// ExpectNonNilErr runs the given statement and verifies that it returns an error.
+func (sr *SQLRunner) ExpectNonNilErr(t Fataler, query string, args ...interface{}) {
+	helperOrNoop(t)()
+	_, err := sr.DB.ExecContext(context.Background(), query, args...)
+	if err == nil {
+		t.Fatalf("expected query '%s' to return a non-nil error", query)
+	}
+}
+
 func (sr *SQLRunner) expectErr(t Fataler, query string, err error, errRE string) {
 	helperOrNoop(t)()
 	if !testutils.IsError(err, errRE) {

